lib/pacemaker/pcmk_status.c

```diff
@@ -70,76 +70,38 @@ fencing_connect(void)
     }
 }
 
-static void
-pacemakerd_event_cb(pcmk_ipc_api_t *pacemakerd_api,
-                    enum pcmk_ipc_event event_type, crm_exit_t status,
-                    void *event_data, void *user_data)
-{
-    pcmk_pacemakerd_api_reply_t *reply = event_data;
-    enum pcmk_pacemakerd_state *state =
-        (enum pcmk_pacemakerd_state *) user_data;
-
-    /* we are just interested in the latest reply */
-    *state = pcmk_pacemakerd_state_invalid;
-
-    if (event_type != pcmk_ipc_event_reply || status != CRM_EX_OK) {
-        return;
-    }
-
-    if (reply->reply_type == pcmk_pacemakerd_reply_ping &&
-        reply->data.ping.last_good != (time_t) 0 &&
-        reply->data.ping.status == pcmk_rc_ok) {
-        *state = reply->data.ping.state;
-    }
-}
-
-static int
-pacemakerd_status(pcmk__output_t *out)
-{
-    int rc = pcmk_rc_ok;
-    pcmk_ipc_api_t *pacemakerd_api = NULL;
-    enum pcmk_pacemakerd_state state = pcmk_pacemakerd_state_invalid;
-
-    rc = pcmk_new_ipc_api(&pacemakerd_api, pcmk_ipc_pacemakerd);
-    if (pacemakerd_api == NULL) {
-        out->err(out, "Could not connect to pacemakerd: %s",
-                 pcmk_rc_str(rc));
-        return rc;
-    }
-
-    pcmk_register_ipc_callback(pacemakerd_api, pacemakerd_event_cb, (void *) &state);
-
-    rc = pcmk_connect_ipc(pacemakerd_api, pcmk_ipc_dispatch_sync);
```

Comment:
I'm not sure how we should handle the dispatch type. I believe it was sync here because pcmk_status() might be called without a mainloop. pcmk__pacemakerd_status() runs its own mainloop, so that might handle that situation, but I'm not sure that's a good idea if the caller already has one, or how we might integrate with an existing mainloop (which shouldn't end once we get a reply). Hmmm ...

Reply:
Can there only be one at a time globally? Otherwise I don't see the harm in having this second mainloop temporarily that's only reachable from the local `data` object. I'm not going to pretend that I'm well-acquainted with how either GLib's or Pacemaker's main loop code works though, including the idea of a GMainContext.

We could always have a bool or an enum that says which type of dispatch to use, or use sync automatically if the timeout is 0. It would complicate the interface and implementation a little bit, but that might be preferable to reasoning about multiple mainloops if we don't feel confident.

Reply:
There can be more than one mainloop, but one will be the default, and pacemaker attaches various things to the default mainloop. That's just another gotcha to worry about -- we'd have to make sure this mainloop can't be the default, and make sure whatever we do with it doesn't assume the default mainloop is desired.

Using sync with timeout==0 sounds reasonable.

Reply:
Can we do that in `ipc_connect()` (set to sync if timeout == 0), or do we need to do it elsewhere to avoid affecting the public callers' behavior?

Another less pretty option is to pass a dispatch type to `pcmk__pacemakerd_status()`, `pcmk__controller_status()`, `pcmk__designated_controller()`, and `ipc_connect()`. That way for example `pcmk__status()` can choose sync if timeout == 0, but the public `pcmk_pacemakerd_status()` function will still get pcmk_ipc_dispatch_main with default timeout 30000.

Reply:
We can keep public calls the same by passing the default timeout explicitly

Reply:
Oh yeah good point

Reply:
Though it might make sense for public calls to use sync. Not a whole lot of point in running a mainloop that essentially emulates a sync call

Reply:
Yeah, only difference I'm aware of is the 30s vs 5s timeout